# Creating Tables by importing json file

In [1]:
import json
import pandas as pd
import numpy as np

In [2]:
# List of json file names
file_names = ['file2.json', 'file3.json', 'file4.json']

df_list = []  # This list will hold individual dataframes created from each JSON file's 'restaurants' data

# Loop through the file names
for file_name in file_names:
    
    # Load the JSON file
    with open(file_name, 'r') as file:  # Open the JSON file in read mode
        data = json.load(file) 
    
    # Loop through each item in the JSON file
    for item in data:
        
        # Check if the 'restaurants' key exists in the current item
        if 'restaurants' in item:
            
            # Normalize the 'restaurants' data and create a DataFrame
            df_restaurants = pd.json_normalize(item['restaurants'])
            
            # Append the DataFrame to the list
            df_list.append(df_restaurants)

# Concatenate all DataFrames into one
combined_df = pd.concat(df_list, ignore_index=True)

In [3]:
len(combined_df)

28153

In [4]:
combined_df.tail()

restaurant.R.res_id                 restaurant.apikey restaurant.id  \
28148             17557488  9421941dbc9358a88376f2ee33bbb06c      17557488   
28149             18182702  9421941dbc9358a88376f2ee33bbb06c      18182702   
28150             16608864  9421941dbc9358a88376f2ee33bbb06c      16608864   
28151                13231  9421941dbc9358a88376f2ee33bbb06c         13231   
28152             16613649  9421941dbc9358a88376f2ee33bbb06c      16613649   

                restaurant.name  \
28148     Burger Queen Drive In   
28149  Nonna's Pasta & Pizzeria   
28150         Taste of Balingup   
28151                Le Plaisir   
28152        Vivo Bar and Grill   

                                          restaurant.url  \
28148  https://www.zomato.com/lakeview-or/burger-quee...   
28149  https://www.zomato.com/laguna/nonnas-pasta-piz...   
28150  https://www.zomato.com/balingup-wa/taste-of-ba...   
28151  https://www.zomato.com/pune/le-plaisir-deccan-...   
28152  https://www.zomato.com/palm-cove-qld/vivo-bar-...   

                             restaurant.location.address  \
28148                     109 S F St, Lakeview, OR 97630   
28149  Ground Floor, Building G, Solenad 3, Nuvali, D...   
28150                 63 South Western Hwy, Balingup, WA   
28151  759/125, Rajkamal, Opposite Kelkar Eye Hospita...   
28152              49 Williams Esplanade, Palm Cove, QLD   

          restaurant.location.locality restaurant.location.city  \
28148                         Lakeview                 Lakeview   
28149  Solenad 3, Don Jose, Santa Rosa               Santa Rosa   
28150                         Balingup                 Balingup   
28151                  Deccan Gymkhana                     Pune   
28152                        Palm Cove                Palm Cove   

       restaurant.location.city_id restaurant.location.latitude  ...  \
28148                         8397                42.1885000000  ...   
28149                        11071                14.2376789732  ...   
28150                         1924               -33.7845269000  ...   
28151                            5                18.5142099831  ...   
28152                         2430               -16.7480830000  ...   

      restaurant.has_online_delivery restaurant.is_delivering_now  \
28148                              0                            0   
28149                              0                            0   
28150                              0                            0   
28151                              0                            0   
28152                              0                            0   

                restaurant.deeplink restaurant.has_table_booking  \
28148  zomato://restaurant/17557488                            0   
28149  zomato://restaurant/18182702                            0   
28150  zomato://restaurant/16608864                            0   
28151     zomato://restaurant/13231                            0   
28152  zomato://restaurant/16613649                            0   

                                   restaurant.events_url  \
28148  https://www.zomato.com/lakeview-or/burger-quee...   
28149  https://www.zomato.com/laguna/nonnas-pasta-piz...   
28150  https://www.zomato.com/balingup-wa/taste-of-ba...   
28151  https://www.zomato.com/pune/le-plaisir-deccan-...   
28152  https://www.zomato.com/palm-cove-qld/vivo-bar-...   

      restaurant.establishment_types  restaurant.book_url  \
28148                             []                  NaN   
28149                             []                  NaN   
28150                             []                  NaN   
28151                             []                  NaN   
28152                             []                  NaN   

       restaurant.zomato_events restaurant.order_url restaurant.order_deeplink  
28148                       NaN                  NaN                       NaN  
28149                       NaN                  NaN   

In [5]:
combined_df.rename(columns=lambda x: x.replace('restaurant.', ''), inplace=True)
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28153 entries, 0 to 28152
Data columns (total 38 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   R.res_id                      28153 non-null  int64 
 1   apikey                        28153 non-null  object
 2   id                            28153 non-null  object
 3   name                          28153 non-null  object
 4   url                           28153 non-null  object
 5   location.address              28153 non-null  object
 6   location.locality             28153 non-null  object
 7   location.city                 28153 non-null  object
 8   location.city_id              28153 non-null  int64 
 9   location.latitude             28153 non-null  object
 10  location.longitude            28153 non-null  object
 11  location.zipcode              28153 non-null  object
 12  location.country_id           28153 non-null  int64 
 13  location.localit

**Dropping the unwanted columns from the dataframe

In [6]:
df_dropped = combined_df.drop(columns=['R.res_id', 'apikey', 'url', 'offers', 'thumb', 'photos_url', 'menu_url', 
                                       'featured_image', 'deeplink', 'events_url', 'establishment_types', 'book_url',
                                      'zomato_events', 'order_url', 'order_deeplink'])
df_dropped.tail()

id                      name  \
28148  17557488     Burger Queen Drive In   
28149  18182702  Nonna's Pasta & Pizzeria   
28150  16608864         Taste of Balingup   
28151     13231                Le Plaisir   
28152  16613649        Vivo Bar and Grill   

                                        location.address  \
28148                     109 S F St, Lakeview, OR 97630   
28149  Ground Floor, Building G, Solenad 3, Nuvali, D...   
28150                 63 South Western Hwy, Balingup, WA   
28151  759/125, Rajkamal, Opposite Kelkar Eye Hospita...   
28152              49 Williams Esplanade, Palm Cove, QLD   

                     location.locality location.city  location.city_id  \
28148                         Lakeview      Lakeview              8397   
28149  Solenad 3, Don Jose, Santa Rosa    Santa Rosa             11071   
28150                         Balingup      Balingup              1924   
28151                  Deccan Gymkhana          Pune                 5   
28152                        Palm Cove     Palm Cove              2430   

      location.latitude location.longitude location.zipcode  \
28148     42.1885000000    -120.3458000000            97630   
28149     14.2376789732     121.0565873981                    
28150    -33.7845269000     115.9844924000            06253   
28151     18.5142099831      73.8384293765                0   
28152    -16.7480830000     145.6707680000             4879   

       location.country_id  ... average_cost_for_two  price_range currency  \
28148                  216  ...                   10            1        $   
28149                  162  ...                  850            3        P   
28150                   14  ...                   20            2        $   
28151                    1  ...                 1000            3      Rs.   
28152                   14  ...                   30            3        $   

       user_rating.aggregate_rating  user_rating.rating_text  \
28148                           3.6                     Good   
28149                           4.0                Very Good   
28150                           3.2                  Average   
28151                           4.8                Excellent   
28152                           4.4                Very Good   

      user_rating.rating_color user_rating.votes has_online_delivery  \
28148                   9ACD32                41                   0   
28149                   5BA829                72                   0   
28150                   CDD614                21                   0   
28151                   3F7E00              2510                   0   
28152                   5BA829               381                   0   

      is_delivering_now has_table_booking  
28148                 0                 0  
28149                 0                 0  
28150                 0                 0  
28151                 0                 0  
28152                 0                 0  

[5 rows x 23 columns]

**There are 28151 rows of data after combining 3 json files. Obviously, there are repeated rows.

In [7]:
df_dropped.to_csv('repeated_zomato_data.csv', index=False)

In [8]:
df0 = pd.read_csv('repeated_zomato_data.csv')
df0.tail()

id                      name  \
28148  17557488     Burger Queen Drive In   
28149  18182702  Nonna's Pasta & Pizzeria   
28150  16608864         Taste of Balingup   
28151     13231                Le Plaisir   
28152  16613649        Vivo Bar and Grill   

                                        location.address  \
28148                     109 S F St, Lakeview, OR 97630   
28149  Ground Floor, Building G, Solenad 3, Nuvali, D...   
28150                 63 South Western Hwy, Balingup, WA   
28151  759/125, Rajkamal, Opposite Kelkar Eye Hospita...   
28152              49 Williams Esplanade, Palm Cove, QLD   

                     location.locality location.city  location.city_id  \
28148                         Lakeview      Lakeview              8397   
28149  Solenad 3, Don Jose, Santa Rosa    Santa Rosa             11071   
28150                         Balingup      Balingup              1924   
28151                  Deccan Gymkhana          Pune                 5   
28152                        Palm Cove     Palm Cove              2430   

       location.latitude  location.longitude location.zipcode  \
28148          42.188500         -120.345800            97630   
28149          14.237679          121.056587              NaN   
28150         -33.784527          115.984492            06253   
28151          18.514210           73.838429                0   
28152         -16.748083          145.670768             4879   

       location.country_id  ... average_cost_for_two  price_range currency  \
28148                  216  ...                   10            1        $   
28149                  162  ...                  850            3        P   
28150                   14  ...                   20            2        $   
28151                    1  ...                 1000            3      Rs.   
28152                   14  ...                   30            3        $   

       user_rating.aggregate_rating  user_rating.rating_text  \
28148                           3.6                     Good   
28149                           4.0                Very Good   
28150                           3.2                  Average   
28151                           4.8                Excellent   
28152                           4.4                Very Good   

      user_rating.rating_color  user_rating.votes has_online_delivery  \
28148                   9ACD32                 41                   0   
28149                   5BA829                 72                   0   
28150                   CDD614                 21                   0   
28151                   3F7E00               2510                   0   
28152                   5BA829                381                   0   

      is_delivering_now  has_table_booking  
28148                 0                  0  
28149                 0                  0  
28150                 0                  0  
28151                 0                  0  
28152                 0                  0  

[5 rows x 23 columns]

**Checking the number of times a specific id that has been repeated.

In [9]:
# ID to check the amount of its recurrence
specific_restaurant_id = 17557488

# Filter the DataFrame for rows with the specified restaurant_id
filtered_df = df0[df0['id'] == specific_restaurant_id]

# Get the count of rows
num_rows = len(filtered_df)
num_rows

182

In [11]:
df_unique = df0.drop_duplicates(keep='first')

**Checking the number of times a specific id that has been repeated after dropping the duplicates

In [12]:
# Again verifying the recurrence of the same id
specific_restaurant_id = 17557488

# Filter the DataFrame for rows with the specified restaurant_id
filtered_df = df_unique[df_unique['id'] == specific_restaurant_id]

# Get the count of rows
num_rows = len(filtered_df)
num_rows

1

In [13]:
df_unique.to_csv('non_repeated.csv', index=False)

In [14]:
df = pd.read_csv('non_repeated.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8004 entries, 0 to 8003
Data columns (total 23 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   id                            8004 non-null   int64  
 1   name                          8004 non-null   object 
 2   location.address              8004 non-null   object 
 3   location.locality             8004 non-null   object 
 4   location.city                 8004 non-null   object 
 5   location.city_id              8004 non-null   int64  
 6   location.latitude             8004 non-null   float64
 7   location.longitude            8004 non-null   float64
 8   location.zipcode              2598 non-null   object 
 9   location.country_id           8004 non-null   int64  
 10  location.locality_verbose     8004 non-null   object 
 11  switch_to_order_menu          8004 non-null   int64  
 12  cuisines                      8003 non-null   object 
 13  ave

**The column 'cuisines' has multiple values in it suggesting the bigger dataframe/table is not even in 1NF.
**Splitting the cuisine so that each row contains only one cuisine name.

In [15]:
cuisine_df = df['cuisines'].str.split(',', expand=True)

# Create a DataFrame with cuisine IDs and names
cuisine_id_df = pd.DataFrame({
    'cuisine_name': cuisine_df.stack().unique()
})

cuisine_id_df.tail()

cuisine_name
164        Filipino
165        Bar Food
166  South American
167      Portuguese
168           Greek

**Assigning random cuisine_id for each cuisine name.

In [16]:
cuisine_id_df['cuisine_id'] = np.random.randint(1111, 9999, size=len(cuisine_id_df))

cuisine_id_df.head()

cuisine_name  cuisine_id
0           Asian        7456
1            Cafe        8484
2         Italian        3914
3     Continental        3876
4  Coffee and Tea        3603

**Adding 'id' of the restaurant to the cuisine dataframe which can act like foreign key referencing the restaurant table/dataframe

In [17]:
# Create a mapping between restaurant IDs and cuisine IDs
restaurant_id_df = df[['id']].copy()

# Expand the DataFrame with multiple cuisine values into separate rows
expanded_cuisine_df = pd.DataFrame({
    'id': df['id'].repeat(cuisine_df.shape[1]),
    'cuisine_name': cuisine_df.values.flatten()
})

# Merge the DataFrame with cuisine IDs with the expanded DataFrame based on the cuisine names
cuisine_table_df = pd.merge(expanded_cuisine_df, cuisine_id_df, on='cuisine_name')

cuisine_table_df = pd.merge(cuisine_table_df, restaurant_id_df, on='id')

cuisine_table_df.tail()

id    cuisine_name  cuisine_id
16148  18365575      Vietnamese        6509
16149      7814           Oriya        9263
16150      9646           Oriya        9263
16151  18463567  South American        3188
16152      2443  South American        3188

In [18]:
cuisine_table_df.to_csv('cuisine.csv', index=False)

**Creating location table from the original dataframe keeping in mind to keep city_id as the primary key
**'id' and 'country_id' can be specified as foreign key referencing respecting table

In [19]:
location_df = df[['id', 'location.address', 'location.locality', 'location.city', 'location.latitude', 
                  'location.longitude', 'location.zipcode', 'location.country_id', 'location.locality_verbose', 
                  'location.city_id']].copy()

# Remove 'location.' prefix from the column names
location_df.columns = location_df.columns.str.replace('location.', '')

# Display the location DataFrame
location_df.tail()

C:\Users\robin\AppData\Local\Temp\ipykernel_15268\4210274601.py:6: FutureWarning: The default value of regex will change from True to False in a future version.
  location_df.columns = location_df.columns.str.replace('location.', '')


id                                            address    locality  \
7999  18273942  F-7, Gopalji Mart, Near Jaypee Hospital, Secto...  Sector 125   
8000    311334  Mahagun Modern, Sector 78, Near, Sector 72, Noida   Sector 72   
8001  18424173    Main Market, Near ICICI Bank, Sector 110, Noida  Sector 110   
8002  18424175             A 2/22, Main Market, Sector 110, Noida  Sector 110   
8003  18275704                  Location Varies, Sector 37, Noida   Sector 37   

       city   latitude  longitude zipcode  country_id   locality_verbose  \
7999  Noida  28.516344  77.371427     NaN           1  Sector 125, Noida   
8000  Noida  28.515425  77.373336     NaN           1   Sector 72, Noida   
8001  Noida  28.533741  77.387546     NaN           1  Sector 110, Noida   
8002  Noida  28.533194  77.387116     NaN           1  Sector 110, Noida   
8003  Noida  28.529984  77.375657     NaN           1   Sector 37, Noida   

      city_id  
7999        1  
8000        1  
8001        1  
8002        1  
8003        1

In [20]:
location_df.to_csv('location.csv', index=False)

In [21]:
ratings_df = df[['user_rating.aggregate_rating', 'user_rating.rating_text',
                 'user_rating.rating_color']].copy()

# Remove prefix from the column names
ratings_df.columns = ratings_df.columns.str.replace('user_rating.', '')

ratings_df.tail()

C:\Users\robin\AppData\Local\Temp\ipykernel_15268\4130328020.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  ratings_df.columns = ratings_df.columns.str.replace('user_rating.', '')


aggregate_rating rating_text rating_color
7999               3.9        Good       9ACD32
8000               2.1        Poor       FF7800
8001               0.0   Not rated       CBCBC8
8002               0.0   Not rated       CBCBC8
8003               3.6        Good       9ACD32

**Noticed that there are 6 values for 'rating_text' and 'rating_color'
**Made a small dataframe with these values and assigned 'rating_id' for them

In [22]:
rating_text = ['Excellent', 'Very Good', 'Good', 'Average', 'Poor', 'Not rated']
rating_color = ['3F7E00', '5BA829', '9ACD32', 'CDD614', 'FF7800', 'CBCBC8']

# Create a new DataFrame
new_ratings_df = pd.DataFrame({
    'rating_text': rating_text,
    'rating_color': rating_color
})

# Add a new column 'rating_id' with randomly generated 5-digit numbers
new_ratings_df['rating_id'] = np.random.randint(111, 999, size=len(new_ratings_df))

new_ratings_df.tail(6)

rating_text rating_color  rating_id
0   Excellent       3F7E00        275
1   Very Good       5BA829        932
2        Good       9ACD32        243
3     Average       CDD614        975
4        Poor       FF7800        737
5   Not rated       CBCBC8        916

In [23]:
new_ratings_df.to_csv('ratings.csv', index=False)

In [24]:
services_df = df[['id', 'has_online_delivery', 'is_delivering_now', 'has_table_booking']].copy()

services_df.tail()

id  has_online_delivery  is_delivering_now  has_table_booking
7999  18273942                    1                  0                  0
8000    311334                    1                  0                  0
8001  18424173                    0                  0                  0
8002  18424175                    0                  0                  0
8003  18275704                    1                  0                  0

**The columns 'has_online_deliverey', 'is_delivering_now', 'has_table_booking' can have only 8 combinations of 0s and 1s.
**made a small dataframe 'services_df' with random generated services_id

In [25]:
# Generate services_id randomly in the range of 300-400
np.random.seed(42)
services_df = pd.DataFrame({
    'services_id': np.random.randint(300, 401, size=8),
    'has_online_delivery': [0, 0, 0, 0, 1, 1, 1, 1],
    'is_delivering_now': [0, 0, 1, 1, 0, 0, 1, 1],
    'has_table_booking': [0, 1, 0, 1, 0, 1, 0, 1]
})

services_df

services_id  has_online_delivery  is_delivering_now  has_table_booking
0          351                    0                  0                  0
1          392                    0                  0                  1
2          314                    0                  1                  0
3          371                    0                  1                  1
4          360                    1                  0                  0
5          320                    1                  0                  1
6          382                    1                  1                  0
7          386                    1                  1                  1

In [26]:
services_df.to_csv('services.csv', index=False)

**Making 'restaurant_df' dataframe with 'rating_id', 'services_id' mapped from corresponding tables.

In [27]:
rating_mapping = dict(zip(new_ratings_df['rating_text'], new_ratings_df['rating_id']))

# Map rating_text to rating_id using the mapping dictionary
df['rating_id'] = df['user_rating.rating_text'].map(rating_mapping)

restaurant_df = df[['id', 'name', 'switch_to_order_menu', 'average_cost_for_two', 'price_range', 
                                      'currency', 'user_rating.aggregate_rating', 'user_rating.votes', 'rating_id']].copy()

restaurant_df.columns = restaurant_df.columns.str.replace('user_rating.', '')

# Merge services_df with main_df to get the service_id for each combination of 0s and 1s
services_id_combination = pd.merge(df, services_df, 
                                   on=['has_online_delivery', 'is_delivering_now', 'has_table_booking'], 
                                   how='left')

restaurant_df = pd.merge(restaurant_df, services_id_combination[['id', 'services_id']], 
                                      on='id', 
                                      how='left')

restaurant_df.head()

C:\Users\robin\AppData\Local\Temp\ipykernel_15268\3124375287.py:9: FutureWarning: The default value of regex will change from True to False in a future version.
  restaurant_df.columns = restaurant_df.columns.str.replace('user_rating.', '')


id                     name  switch_to_order_menu  \
0  16668008            Arigato Sushi                     0   
1    801690                    Mocha                     0   
2  17558738          Blue House Cafe                     0   
3  16611701              Star Buffet                     0   
4   2100784  11th Avenue Cafe Bistro                     0   

   average_cost_for_two  price_range currency  aggregate_rating  votes  \
0                    25            2        $               3.3     26   
1                   800            3      Rs.               4.6    567   
2                    10            1        $               4.3     88   
3                    20            2        $               2.9     11   
4                   400            2      Rs.               4.1    377   

   rating_id  services_id  
0        975          351  
1        275          351  
2        932          351  
3        975          351  
4        932          351

In [28]:
restaurant_df.to_csv('restaurant.csv', index=False)